asserts/user.go

```diff
@@ -0,0 +1,262 @@
+func checkHashedPassword(headers map[string]interface{}, name string) (string, error) {
+	pw, err := checkOptionalString(headers, name)
+	if err != nil {
+		return "", err
+	}
+	// the pw string is optional, so just return if its empty
+	if pw == "" {
+		return "", nil
+	}
+
+	// parse the shadow line
+	shd, err := parseShadowLine(pw)
+	if err != nil {
+		return "", fmt.Errorf(`%q header invalid: %s`, name, err)
+	}
+
+	// and verify it
+
+	// see crypt(3), ID 6 means SHA-512 (since glibc 2.7)
+	ID, err := strconv.Atoi(shd.ID)
+	if err != nil {
+		return "", fmt.Errorf(`%q header must start with "$integer-id$", got %q`, name, shd.ID)
+	}
+	// double check that we only allow modern hashes
+	if ID < 6 {
+		return "", fmt.Errorf("%q header only supports $id$ values of 6 (sha512crypt) or higher", name)
+	}
+
+	// the $rounds=N$ part is optional
+	if strings.HasPrefix(shd.Rounds, "rounds=") {
+		rounds, err := strconv.Atoi(strings.SplitN(shd.Rounds, "=", 2)[1])
+		if err != nil {
+			return "", fmt.Errorf("%q header has invalid number of rounds: %s", name, err)
+		}
+		if rounds < 5000 || rounds > 999999999 {
+			return "", fmt.Errorf("%q header rounds parameter out of bounds: %d", name, rounds)
+		}
+	}
+
+	// see crypt(3) for the legal chars
+	validSaltAndHash := regexp.MustCompile(`^[a-zA-Z0-9./]+$`)
+	if !validSaltAndHash.MatchString(shd.Salt) {
+		return "", fmt.Errorf("%q header has invalid chars in salt %q", name, shd.Salt)
+	}
+	if !validSaltAndHash.MatchString(shd.Hash) {
+		return "", fmt.Errorf("%q header has invalid chars in hash %q", name, shd.Hash)
+	}
+
+	return pw, nil
+}
+
+func assembleSystemUser(assert assertionBase) (Assertion, error) {
+	err := checkAuthorityMatchesBrand(&assert)
+	if err != nil {
+		return nil, err
+	}
+	email, err := checkNotEmptyString(assert.headers, "email")
+	if err != nil {
+		return nil, err
+	}
+	if _, err := mail.ParseAddress(email); err != nil {
+		return nil, fmt.Errorf(`"email" header must be a RFC 5322 compliant email address: %s`, err)
+	}
+
+	series, err := checkStringList(assert.headers, "series")
+	if err != nil {
+		return nil, err
+	}
+	models, err := checkStringList(assert.headers, "models")
+	if err != nil {
+		return nil, err
+	}
+	if _, err := checkOptionalString(assert.headers, "name"); err != nil {
+		return nil, err
+	}
+	if _, err := checkStringMatches(assert.headers, "username", validSystemUserUsernames); err != nil {
+		return nil, err
+	}
+	if _, err := checkHashedPassword(assert.headers, "password"); err != nil {
+		return nil, err
+	}
+
+	sshKeys, err := checkStringList(assert.headers, "ssh-keys")
+	if err != nil {
+		return nil, err
+	}
+	since, err := checkRFC3339Date(assert.headers, "since")
+	if err != nil {
+		return nil, err
+	}
+	until, err := checkRFC3339Date(assert.headers, "until")
+	if err != nil {
+		return nil, err
+	}
+	if until.Before(since) {
+		return nil, fmt.Errorf("'until' time cannot be before 'since' time")
+	}
+	if until.After(time.Now().AddDate(1, 0, 0)) {
```

Comment:
assemble is used also when we parse, so this cannot be expressed that way otherwise after a year we cannot even parse the assertion,  we can only check the difference from since